### DISCUSSION TITLE
h11._util.LocalProtocolError issue

### DISCUSSION THREAD

**Discussion by rshaliamekh** [fastapi/fastapi#9921] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
@app.middleware("http")
async def my_middleware(request: Request, call_next):
    response = await call_next(request)
    response.headers["X-My-Header"] = "test"
    return response
```

### Description

Hello everyone! For our project we use FastAPI for a custom made api-gateway micro service that redirects the requests with aiohttp to other services. The app is dockerized and run in a k8s cluster.
When performing load testing with Locust (specific test configuration: Number of users (peak concurrency) - 500, Spawn rate (users started/second) - 20) on several requests we experience the following error:
```
  File "/usr/local/lib/python3.11/site-packages/starlette/middleware/cors.py", line 83, in __call__
    await self.app(scope, receive, send)
  File "/usr/local/lib/python3.11/site-packages/starlette/middleware/base.py", line 109, in __call__
    await response(scope, receive, send)
  File "/usr/local/lib/python3.11/site-packages/starlette/responses.py", line 270, in __call__
    async with anyio.create_task_group() as task_group:
  File "/usr/local/lib/python3.11/site-packages/anyio/_backends/_asyncio.py", line 597, in __aexit__
    raise exceptions[0]
  File "/usr/local/lib/python3.11/site-packages/starlette/responses.py", line 273, in wrap
    await func()
  File "/usr/local/lib/python3.11/site-packages/starlette/middleware/base.py", line 134, in stream_response
    return await super().stream_response(send)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/starlette/responses.py", line 255, in stream_response
    await send(
  File "/usr/local/lib/python3.11/site-packages/starlette/middleware/errors.py", line 159, in _send
    await send(message)
  File "/usr/local/lib/python3.11/site-packages/uvicorn/protocols/http/h11_impl.py", line 512, in send
    output = self.conn.send(event)
             ^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/h11/_connection.py", line 512, in send
    data_list = self.send_with_data_passthrough(event)
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/h11/_connection.py", line 527, in send_with_data_passthrough
    raise LocalProtocolError("Can't send data when our state is ERROR")
h11._util.LocalProtocolError: Can't send data when our state is ERROR
```
After debugging the issue we found out that the error was caused by our custom but very basic middleware, using which we just set some specific response headers:
```
@app.middleware("http")
async def my_middleware(request: Request, call_next):
    response = await call_next(request)
    response.headers["X-My-Header"] = "test"
    return response
```
Beside this we also use `starlette.middleware.cors.CORSMiddleware` and `uvicorn.middleware.proxy_headers.proxyHeadersMiddleware`, but as we identified they were not the source of the issue. 

To resolve the issue we replaced the middleware shown above with this one:
```
from starlette.datastructures import MutableHeaders
from starlette.middleware.base import BaseHTTPMiddleware
from starlette.types import ASGIApp, Message, Scope, Send


class MyMiddleware:
    def __init__(self, app: ASGIApp, headers: dict[str, str]):
        self.app = app
        self.custom_headers = headers

    async def __call__(self, scope: Scope, receive: ASGIApp, send: Send) -> None:
        async def send_wrapper(message: Message) -> None:
            if not message.get("headers"):
                message["headers"] = []
            headers = MutableHeaders(scope=message)
            for k, v in self.custom_headers.items():
                headers.append(k, v)
            await send(message)

        await self.app(scope, receive, send_wrapper)

app.add_middleware(AdditionalHeadersMiddleware, headers={"X-My-Header": "test"})
```
Unfortunately we could not reproduce the issue locally.
Any idea what may cause this error?

### Operating System

Linux

### FastAPI Version

0.97.0

### Python Version

3.11.4

**Comment by Kludex** [comment, marked as answer]
I believe the problem is on `uvicorn`. It will be cool to check if there's an issue on `uvicorn`'s side, or a discussion about it, and if not, create a discussion.

To avoid the problem, you can install `httptools`: `pip install httptools`.

**Comment by RameshByndoor** [comment]
Are there any further development or any hints. I am facing similar issue when concurrent users are increased. @rshaliamekh were you able to figure out solve this?
